### ISSUE TITLE
Let the serializer be set programmatically!

### ISSUE THREAD

**Issue by lamba92** [Litote/kmongo#243]
When initializing a client, there is no possibility to choose the serialization module or configure it. Probably this is a choice made to let KMongo figure out at runtime which serialization engine is available and pick one from the classpath using reflections or some other _Java-old-school-black-magic_.

For example Ktor allows to explicitely set a `Json` formatter when initializing the module (docs [here](https://ktor.io/docs/serialization-converter.html#configuration)). The equivalent in KMongo would be to create an interface that wraps the various implementations of serializers and provide defaults implementations of those serialization interfaces for the engines already implemented and eventually let the developer choose which engine to use not by classpath black magic but by explicitely calling the right serializer like:
```kotlin
interface KMongoSerializer {
    fun <T : Any> serialize(obj: T, type: KType): Bson
    fun <T : Any> deserialize(bson: Bson, type: KType): T
}

@OptIn(ExperimentalStdlibApi::class)
inline fun <reified T: Any> KMongoSerializer.serialize(obj: T) =
    serialize(obj, typeOf<T>())

@OptIn(ExperimentalStdlibApi::class)
inline fun <reified T: Any> KMongoSerializer.deserialize(bson: Bson) =
    deserialize<T>(bson, typeOf<T>())

class KotlinxSerializer(private val kbson: KBson) : KMongoSerializer {
    override fun <T : Any> serialize(obj: T, type: KType): Bson {
        // use this.kbson to deserialialize
    }
    override fun <T : Any> deserialize(bson: Bson, type: KType): T {
        // use this.kbson to serialize
    }
}

class KMongoDBClient(client: MongoClient, serializer: KMongoSerializer) {
    // actual implementation and dispatch of calls only for this client
}

fun main() {
    val serializer: KotlinxSerializer =  // whatever with custom contextual stuff ecc...
    KMongo.createClient("localhost:1234", serializer)
}
```
Have a look on how kotlinx.serialization `Json` recovers a specific serializer using its `KType` [here](https://github.com/Kotlin/kotlinx.serialization/blob/master/core/commonMain/src/kotlinx/serialization/Serializers.kt#L31-L33).

Also a programmatic approach for choosing a serializer would be easier to for developer to use, more concise and explicit and Kotlin friendly. I also believe it would be easier for maintain the KMongo codebase.

**Comment by zigzago** [comment]
Hello, 

As you might expect, there is a  [ClassMappingTypeService](http://litote.org/kmongo/dokka/kmongo/org.litote.kmongo.service/-class-mapping-type-service/) interface that is loaded by the [ServiceLoader](https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/util/ServiceLoader.html) java mechanism at runtime.

If there is more than one implementation in the classpath, the [priority](http://litote.org/kmongo/dokka/kmongo/org.litote.kmongo.service/-class-mapping-type-service/priority.html) property of each `ClassMappingTypeService` is used to select the service. So you can already select your mapping engine by providing your own implementation (expert usage ;) ).

Do you have a use case with two mapping engines for KMongo in the same application ?

**Comment by lamba92** [comment]
In my application I have `Json`, `KBson` and `Protobuf` serializers and while `Json` and `Protobuf` shares the same `SerializersModule` that is contained in my dependency injection container, `KBson` is built by the `ServiceLoader`. 

Indeed I could write a serializer by myself for KMongo but in my use case I just need to add contextuals to the `KBson` `serializersModule`. Also we plan to be able to serialize differently some classes at runtime and this would not be possible with the actual implementation of the KMong serialization loading mechanism.

To summarize, I need to easily pass a custom `SerializerModule` to the `KBson` that KMongo will use and it would be ideal to do it it without runtime magic. This solution is needed because the data classes I am trying to serialize do not have their serializer in their project!

**Comment by zigzago** [comment]
You can already register a custom module or serializer using `registerModule` or `registerSerializer` (look at https://litote.org/kmongo/object-mapping/#additional-modules-and-serializers )

Does it feed your need ?
